### PR TITLE
[FAI-2451] Workday Source Connector: Custom Report Support

### PR DESCRIPTION
**Pull request by omreego**
## Description
Adding Custom Report Support

## Type of change
- [ ] Bug fix
- [x] New feature
- [ ] Breaking change

### CODE REVIEW

File: sources/workday-source/src/streams/custom_reports.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  getJsonSchema(): Dictionary<any, string> {
+    return require('../../resources/schemas/people.json');
+  }
+  get primaryKey(): StreamKey {
+    return ['id'];
+  }
+
+  async *readRecords(): AsyncGenerator<Person> {
+    const workday = await Workday.instance(this.cfg, this.logger);
+    yield* workday.people();
```
Comment (tovbinm) [inline sources/workday-source/src/streams/custom_reports.ts:25]: `workday.customReport(this.cfg.custom_report_path)`

File: sources/workday-source/src/streams/custom_reports.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  }
+
+  getJsonSchema(): Dictionary<any, string> {
+    return require('../../resources/schemas/people.json');
+  }
+  get primaryKey(): StreamKey {
+    return ['id'];
+  }
+
+  async *readRecords(): AsyncGenerator<Person> {
```
Comment (tovbinm) [inline sources/workday-source/src/streams/custom_reports.ts:23]: `AsyncGenerator<Dictionary<any>>`

File: sources/workday-source/src/streams/custom_reports.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  }
+
+  getJsonSchema(): Dictionary<any, string> {
+    return require('../../resources/schemas/people.json');
+  }
+  get primaryKey(): StreamKey {
+    return ['id'];
+  }
+
+  async *readRecords(): AsyncGenerator<Dictionary<any>> {
```
Comment (tovbinm) [inline sources/workday-source/src/streams/custom_reports.ts:23]: you need to invoke the call here - https://github.com/faros-ai/airbyte-connectors/blob/omree/implementing_raas/sources/workday-source/src/index.ts#L47-L54

File: sources/workday-source/src/streams/custom_reports.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
+import {AirbyteLogger, AirbyteStreamBase, StreamKey} from 'faros-airbyte-cdk';
+import {Dictionary} from 'ts-essentials';
+
+import {WorkdayConfig} from '..';
+import {Person} from '../types';
+import {Workday} from '../workday';
+
+export class CustomReports extends AirbyteStreamBase {
```
Comment (tovbinm) [inline sources/workday-source/src/streams/custom_reports.ts:8]: You need to create the class here - https://github.com/faros-ai/airbyte-connectors/blob/omree/implementing_raas/sources/workday-source/src/index.ts#L47-L54

File: sources/workday-source/src/workday.ts [hunk 5]
```diff
@@ -71,7 +73,8 @@ export class Workday {
       logger,
       api,
       cfg.limit ?? DEFAULT_PAGE_LIMIT,
-      apiBaseUrlTemplate
+      apiBaseUrlTemplate,
```
Comment (tovbinm) [inline sources/workday-source/src/workday.ts:75]: Let's also add `maxBodyLength: Infinity` in `axios.create` call

File: sources/workday-source/src/workday.ts [hunk 6]
```diff
@@ -71,7 +73,8 @@ export class Workday {
...
+      cfg.customReportPath ?? ''
```
Comment (tovbinm) [inline sources/workday-source/src/workday.ts:76]: And lets make the `timeout: 30000` configurable with default of `60000` (similarly to https://github.com/faros-ai/airbyte-connectors/pull/1122)

File: sources/workday-source/src/workday.ts [hunk 7]
```diff
@@ -32,7 +32,8 @@ export class Workday {
     private readonly logger: AirbyteLogger,
     private readonly api: AxiosInstance,
     private readonly limit: number,
-    private readonly apiBaseUrlTemplate: string
+    private readonly apiBaseUrlTemplate: string,
```
Comment (tovbinm) [inline sources/workday-source/src/workday.ts:35]: let's remove `const DEFAULT_BASE_URL` and  https://github.com/faros-ai/airbyte-connectors/blob/omree/implementing_raas/sources/workday-source/src/workday.ts#L56 and instead do: 
```
if (!cfg.baseUrl) {
      throw new VError('baseUrl must not be an empty string');
}
const baseURL = new URL(cfg.baseUrl);
```

File: sources/workday-source/src/workday.ts [hunk 8]
```diff
@@ -51,16 +52,21 @@ export class Workday {
...
+      throw new VError('baseUrl must not be an empty string');
+    }
 
-    const baseURL = new URL(cfg.baseUrl ?? DEFAULT_BASE_URL);
+    const baseUrl = new URL(cfg.baseUrl);
     const apiBaseUrlTemplate =
-      baseURL.toString() + `/api/${VERSION_PLACEHOLDER}/${cfg.tenant}`;
+      baseUrl.toString() + `/api/${VERSION_PLACEHOLDER}/${cfg.tenant}`;
     logger.debug('Assuming API base url template: %s', apiBaseUrlTemplate);
+    const timeout = cfg.timeout ?? 6000;
```
Comment (tovbinm) [inline sources/workday-source/src/workday.ts:63]: 60000

File: sources/workday-source/src/workday.ts [hunk 9]
```diff
@@ -143,6 +151,17 @@ export class Workday {
     );
   }
 
+  async *customReports(path: string): AsyncGenerator<any> {
+    // Note input param path should start with '/'
+    const baseURL = this.baseUrl;
+    const res = await this.api.get(path, {
```
Comment (tovbinm) [inline sources/workday-source/src/workday.ts:157]: Make sure to prepend the /service/customreport2/tenant/ to the path (see my comment in spec.json)

File: sources/workday-source/resources/spec.json [hunk 10]
```diff
@@ -53,8 +53,8 @@
         "type": "number",
         "title": "Request Timeout",
         "minimum": 500,
-        "maximum": 40000,
-        "default": 6000,
+        "maximum": 100000,
```
Comment (tovbinm) [inline sources/workday-source/resources/spec.json:56]: why set the maximum?

Comment (omreego) [reply sources/workday-source/resources/spec.json:56]: No need - I thought it was required

Comment (tovbinm) [reply sources/workday-source/resources/spec.json:56]: Please remove it

File: sources/workday-source/src/workday.ts (hunk 9 above)
Comment (omreego) [reply sources/workday-source/src/workday.ts:157]: I see - adding it now

File: sources/workday-source/src/workday.ts [hunk 11]
```diff
@@ -143,6 +151,17 @@ export class Workday {
     );
   }
 
+  async *customReports(path: string): AsyncGenerator<any> {
+    // Note input param path should start with '/'
+    const baseURL = this.baseUrl + '/service/customreport2';
+    const res = await this.api.get(path, {
+      baseURL,
```
Comment (tovbinm) [inline sources/workday-source/src/workday.ts:158]: Let's add a log line here of the final url we are invoking

File: sources/workday-source/src/workday.ts [hunk 12]
```diff
@@ -55,6 +56,9 @@ export class Workday {
     if (!cfg.baseUrl) {
       throw new VError('baseUrl must not be an empty string');
     }
+    if (!cfg.tenant) {
+      throw new VError('tenant must not be an empty string');
```
Comment (tovbinm) [inline sources/workday-source/src/workday.ts:60]: it is already checked on line 45

File: sources/workday-source/test/index.test.ts [hunk 13]
```diff
@@ -18,6 +18,17 @@ function readTestResourceFile(fileName: string): any {
   return JSON.parse(fs.readFileSync(`test_files/${fileName}`, 'utf8'));
 }
 
+function getWorkdayInstance(logger, axios_instance, limit): Workday {
+  return new Workday(
+    logger,
+    axios_instance,
+    limit,
+    'base-url',
+    'base-url',
```
Comment (tovbinm) [inline sources/workday-source/test/index.test.ts:27]: why base-url repeated twice?

File: sources/workday-source/src/workday.ts [hunk 14]
```diff
@@ -153,7 +158,8 @@ export class Workday {
 
   async *customReports(path: string): AsyncGenerator<any> {
     // Note input param path should start with '/'
-    const baseURL = this.baseUrl + '/service/customreport2';
+    const baseURL = `${this.baseUrl}/service/customreport2/${this.tenant}`;
+    this.logger.info(`base url: ${baseURL}`);
     const res = await this.api.get(path, {
```
Comment (tovbinm) [inline sources/workday-source/src/workday.ts:163]: 1. construct the full url, log it and then invoke it
2. add `format` query param with `json` value

**Review by tovbinm — APPROVED** [review]

**Pull request merged**